server/connection.ts

```diff
@@ -0,0 +1,280 @@
+    this.socket = socket;
+    this.room = room;
+    this.room.roster.push({ id: socket.id });
+    this.room.io.emit('roster', this.room.roster);
+    this.emitInitEvents();
+    this.setupSocketListeners();
+  }
+
+  emitInitEvents() {
+    this.socket.emit('REC:host', this.room.getHostState());
+    this.socket.emit('REC:nameMap', this.room.nameMap);
+    this.socket.emit('REC:pictureMap', this.room.pictureMap);
+    this.socket.emit('REC:tsMap', this.room.tsMap);
+    this.socket.emit('chatinit', this.room.chat);
+    this.socket.emit('playlistUpdate', this.room.videoPlaylist);
+  }
+
+  setupSocketListeners() {
+    this.socket.on('CMD:name', this.setName);
+    this.socket.on('CMD:picture', this.addPicture);
+    this.socket.on('CMD:host', this.host);
+    this.socket.on('CMD:addVideoToPlaylist', this.addVideoToPlaylist);
+    this.socket.on('CMD:play', this.playVideo);
+    this.socket.on('CMD:pause', this.pauseVideo);
+    this.socket.on('CMD:seek', this.seekVideo);
+    this.socket.on('CMD:ts', this.sendVideoTimestamp);
+    this.socket.on('CMD:chat', this.sendChatMessage);
+    this.socket.on('CMD:joinVideo', this.joinVideo);
+    this.socket.on('CMD:leaveVideo', this.leaveVideo);
+    this.socket.on('CMD:joinScreenShare', this.joinScreenShare);
+    this.socket.on('CMD:leaveScreenShare', this.leaveScreenShare);
+    this.socket.on('CMD:startVBrowser', this.startVBrowser);
+    this.socket.on('CMD:stopVBrowser', this.stopVBrowser);
+    this.socket.on('CMD:changeController', this.changeController);
+    this.socket.on('CMD:askHost', this.askHost);
+    this.socket.on('signal', this.emitSignal);
+    this.socket.on('signalSS', this.emitSSSignal);
+    this.socket.on('disconnect', this.disconnect);
+  }
+
+  setName = (data: string) => {
+    if (!data || !this.socket) {
+      return;
+    }
+    if (data && data.length > 100) {
+      return;
+    }
+    this.room.nameMap[this.socket.id] = data;
+    this.room.io.of(this.room.roomId).emit('REC:nameMap', this.room.nameMap);
+  };
+
+  addPicture = (data: string) => {
+    if (data && data.length > 10000) {
+      return;
+    }
+    this.room.pictureMap[this.socket.id] = data;
+    this.room.io
+      .of(this.room.roomId)
+      .emit('REC:pictureMap', this.room.pictureMap);
+  };
+
+  host = (data: PlaylistVideo) => {
+    const sharer = this.room.roster.find((user) => user.isScreenShare);
+    if (sharer) {
+      // Can't update the video while someone is screensharing
+      return;
+    }
+
+    this.room.cmdHost(this.socket, data);
+  };
+
+  addVideoToPlaylist = async (data: string) => {
+    if (!data) {
+      return;
+    }
+
+    const youtubeVideoID = getYoutubeVideoID(data);
+    let video: PlaylistVideo;
+
+    if (youtubeVideoID) {
+      video = await fetchYoutubeVideo(youtubeVideoID);
+    } else {
+      video = {
+        name: data,
+        url: data,
+        img: undefined,
+        duration: 0,
+        durationObject: { h: 0, m: 0, s: 0 },
```

Comment:
what's the purpose of this durationObject? Does it do something that regular duration doesn't do?